File: src/reflection_service/domain.py
import hashlib
from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from uuid import UUID, uuid5

from reflection_service.models import ExtractedClaim, Resolution, ResolutionResult, SearchClaim

SEGMENT_NAMESPACE = UUID("b14d5f2c-4cce-4f37-9e32-ea30c1fd1b42")
ENTITY_NAMESPACE = UUID("966769b5-ab08-4b02-8948-7db052586124")
CLAIM_NAMESPACE = UUID("194eb982-5e16-4011-bdcb-f13a86934da4")


class ExtractionValidationError(ValueError):
    pass


class TerminalExtractionValidationError(ExtractionValidationError):
    pass


def normalize_name(value: str) -> str:
    return " ".join(value.casefold().split())


def segment_id_for(session_id: str, start_user_message_id: str) -> UUID:
    return uuid5(SEGMENT_NAMESPACE, f"{session_id}\0{start_user_message_id}")


def new_entity_id_for(segment_id: UUID, canonical_name: str) -> UUID:
    return uuid5(ENTITY_NAMESPACE, f"{segment_id}\0{normalize_name(canonical_name)}")


def claim_id_for(segment_id: UUID, index: int) -> UUID:
    return uuid5(CLAIM_NAMESPACE, f"{segment_id}\0{index}")


def equivalence_key(
    subject_id: UUID,
    predicate: str,
    *,
    object_entity_id: UUID | None,
    object_value: str | None,
) -> str:
    if (object_entity_id is None) == (object_value is None):
        raise ExtractionValidationError(
            "equivalence key requires exactly one entity object or literal value"
        )
    object_key = (
        f"entity:{object_entity_id}"
        if object_entity_id is not None
        else f"literal:{normalize_name(object_value or '')}"
    )
    raw = f"{subject_id}\0{normalize_name(predicate)}\0{object_key}".encode()
    return hashlib.sha256(raw).hexdigest()


@dataclass(frozen=True, slots=True)
class EntityCandidate:
    id: UUID
    canonical_name: str
    description: str
    aliases: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class MentionContext:
    mention_id: str
    role: str
    text: str
    supporting_claim: str
    candidates: tuple[EntityCandidate, ...]


@dataclass(frozen=True, slots=True)
class PreparedEntity:
    id: UUID
    canonical_name: str
    normalized_name: str
    description: str
    aliases: tuple[str, ...]
    embedding: tuple[float, ...] | None
    is_new: bool


@dataclass(frozen=True, slots=True)
class PreparedClaim:
    id: UUID
    subject: str
    subject_entity_id: UUID
    predicate: str
    confidence: float
    object_entity: str | None
    object_entity_id: UUID | None
    object_value: str | None
    equivalence_key: str
    embedding: tuple[float, ...]


@dataclass(frozen=True, slots=True)
class PreparedSegment:
    id: UUID
    session_id: str
    start_user_message_id: str
    end_user_message_id: str
    summary: str
    entities: tuple[PreparedEntity, ...]
    claims: tuple[PreparedClaim, ...]


@dataclass(frozen=True, slots=True)
class RecallCandidate:
    subject: str
    subject_entity_id: UUID
    predicate: str
    confidence: float
    object_entity: str | None
    object_entity_id: UUID | None
    object_value: str | None
    equivalence_key: str
    segment_id: UUID
    similarity: float
    seed_similarity: float | None
    is_direct: bool


@dataclass(frozen=True, slots=True)
class ClaimSupport:
    segment_ids: list[UUID]
    support_count: int
    session_count: int


def union_candidates(
    trigram: Sequence[EntityCandidate], vector: Sequence[EntityCandidate]
) -> tuple[EntityCandidate, ...]:
    by_id: dict[UUID, EntityCandidate] = {}
    for candidate in (*trigram[:5], *vector[:5]):
        by_id.setdefault(candidate.id, candidate)
    return tuple(by_id.values())


def validate_resolutions(
    contexts: Sequence[MentionContext], result: ResolutionResult
) -> dict[str, tuple[MentionContext, Resolution]]:
    expected = {context.mention_id: context for context in contexts}
    actual = {resolution.mention_id: resolution for resolution in result.resolutions}
    if len(actual) != len(result.resolutions):
        raise ExtractionValidationError("entity resolution returned duplicate mention IDs")
    if actual.keys() != expected.keys():
        raise ExtractionValidationError("entity resolution must return every mention exactly once")
    validated: dict[str, tuple[MentionContext, Resolution]] = {}
    for mention_id, resolution in actual.items():
        allowed = {candidate.id for candidate in expected[mention_id].candidates}
        if (
            resolution.candidate_entity_id is not None
            and resolution.candidate_entity_id not in allowed
        ):
            raise ExtractionValidationError(
                f"entity resolution selected an unknown candidate for {mention_id}"
            )
        validated[mention_id] = (expected[mention_id], resolution)
    return validated


def validate_claim_decisions(
    proposed: Sequence[ExtractedClaim], result: ResolutionResult
) -> list[tuple[int, ExtractedClaim]]:
    expected = {f"c{index}" for index in range(len(proposed))}
    actual = {decision.claim_id: decision for decision in result.claims}
    if len(actual) != len(result.claims):
        raise ExtractionValidationError("claim triage returned duplicate claim IDs")
    if actual.keys() != expected:
        raise ExtractionValidationError(
            "claim triage must return every proposed claim exactly once"
        )

    kept: list[tuple[int, ExtractedClaim]] = []
    for index, claim in enumerate(proposed):
        decision = actual[f"c{index}"]
        if decision.action == "keep":
            kept.append((index, claim))
    return kept


def rank_and_group_claims(
    candidates: Iterable[RecallCandidate],
    support_by_key: dict[str, ClaimSupport],
    limit: int = 20,
) -> list[SearchClaim]:
    grouped: dict[str, tuple[RecallCandidate, float]] = {}
    for candidate in candidates:
        similarity = min(1.0, max(0.0, candidate.similarity))
        confidence = min(1.0, max(0.0, candidate.confidence))
        support = support_by_key.get(candidate.equivalence_key)
        session_count = support.session_count if support is not None else 1
        support_boost = min(0.1, 0.02 * max(0, session_count - 1))
        if candidate.is_direct:
            score = 1.0 + 0.8 * similarity + 0.2 * confidence + support_boost
        else:
            seed_similarity = min(1.0, max(0.0, candidate.seed_similarity or 0.0))
            relevance = 0.5 * similarity + 0.3 * seed_similarity + 0.2 * confidence
            score = 0.8 * relevance + support_boost
        current = grouped.get(candidate.equivalence_key)
        if current is None or score > current[1]:
            grouped[candidate.equivalence_key] = (candidate, score)

    ranked = sorted(grouped.values(), key=lambda item: item[1], reverse=True)[:limit]
    return [
        SearchClaim(
            subject=candidate.subject,
            subject_entity_id=candidate.subject_entity_id,
            predicate=candidate.predicate,
            confidence=candidate.confidence,
            object_entity=candidate.object_entity,
            object_entity_id=candidate.object_entity_id,
            object_value=candidate.object_value,
            segment_ids=(
                support_by_key[candidate.equivalence_key].segment_ids
                if candidate.equivalence_key in support_by_key
                else [candidate.segment_id]
            ),
            support_count=(
                support_by_key[candidate.equivalence_key].support_count
                if candidate.equivalence_key in support_by_key
                else 1
            ),
            session_count=(
                support_by_key[candidate.equivalence_key].session_count
                if candidate.equivalence_key in support_by_key
                else 1
            ),
            score=score,
        )
        for candidate, score in ranked
    ]

File: packages/shared/src/domain.ts
import { createHash } from "node:crypto";

import { caseFold } from "unicode-case-folding";
import { v5 as uuidv5 } from "uuid";

import type {
  ExtractedClaim,
  Resolution,
  ResolutionResult,
  SearchClaim,
  SegmentCreate,
} from "./contracts.js";

export const SEGMENT_NAMESPACE = "b14d5f2c-4cce-4f37-9e32-ea30c1fd1b42";
export const ENTITY_NAMESPACE = "966769b5-ab08-4b02-8948-7db052586124";
export const CLAIM_NAMESPACE = "194eb982-5e16-4011-bdcb-f13a86934da4";

export class ExtractionValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ExtractionValidationError";
  }
}

export class TerminalExtractionValidationError extends ExtractionValidationError {
  constructor(message: string) {
    super(message);
    this.name = "TerminalExtractionValidationError";
  }
}

export function normalizeName(value: string): string {
  return caseFold(value)
    .split(/[\p{White_Space}\u001c-\u001f]+/u)
    .filter(Boolean)
    .join(" ");
}

export function segmentIdFor(
  sessionId: string,
  startUserMessageId: string,
): string {
  return uuidv5(`${sessionId}\0${startUserMessageId}`, SEGMENT_NAMESPACE);
}

function utf8Frame(value: string): string {
  return `${Buffer.byteLength(value, "utf8")}:${value}`;
}

export function segmentIdForRequest(
  request: Pick<
    SegmentCreate,
    | "session_id"
    | "start_user_message_id"
    | "source_boundary_version"
    | "start_source_message_id"
  >,
): string {
  if (request.source_boundary_version === 1) {
    return segmentIdFor(request.session_id, request.start_user_message_id);
  }
  if (request.start_source_message_id === null) {
    throw new Error("V2 segment identity requires a start source cursor");
  }
  return uuidv5(
    `reflection-segment-v2:${utf8Frame(request.session_id)}${utf8Frame(request.start_source_message_id)}`,
    SEGMENT_NAMESPACE,
  );
}

export function sourceFingerprint(request: SegmentCreate): string {
  const messageFrames =
    `${request.messages.length}:` +
    request.messages
      .map((message) => utf8Frame(message.role) + utf8Frame(message.text))
      .join("");
  const source =
    request.source_boundary_version === 1
      ? "reflection-source-v1:" +
        utf8Frame(request.session_id) +
        utf8Frame(request.start_user_message_id) +
        utf8Frame(request.end_user_message_id) +
        messageFrames
      : "reflection-source-v2:" +
        utf8Frame(request.session_id) +
        utf8Frame(request.start_user_message_id) +
        utf8Frame(request.end_user_message_id) +
        utf8Frame(String(request.source_boundary_version)) +
        utf8Frame(request.start_source_message_id) +
        utf8Frame(request.end_source_message_id) +
        messageFrames;
  return createHash("sha256").update(source, "utf8").digest("hex");
}

export function projectionFingerprint(
  segmentId: string,
  endUserMessageId: string,
  summary: string,
  projectionVersion: number,
): string {
  const length = (value: string) => [...value].length;
  const source = `${segmentId}:${length(endUserMessageId)}:${endUserMessageId}:${length(summary)}:${summary}:${projectionVersion}`;
  return createHash("sha256").update(source, "utf8").digest("hex");
}

export type ProjectionFingerprintBoundary =
  | {
      sourceBoundaryVersion: 1;
      endUserMessageId: string;
      endSourceMessageId: null;
    }
  | {
      sourceBoundaryVersion: 2;
      endUserMessageId: string;
      endSourceMessageId: string;
    };

export function projectionFingerprintForBoundary(
  segmentId: string,
  boundary: ProjectionFingerprintBoundary,
  summary: string,
  projectionVersion: number,
): string {
  if (boundary.sourceBoundaryVersion === 1) {
    return projectionFingerprint(
      segmentId,
      boundary.endUserMessageId,
      summary,
      projectionVersion,
    );
  }
  const source =
    "reflection-projection-v2:" +
    utf8Frame(segmentId) +
    utf8Frame(String(boundary.sourceBoundaryVersion)) +
    utf8Frame(boundary.endUserMessageId) +
    utf8Frame(boundary.endSourceMessageId) +
    utf8Frame(summary) +
    utf8Frame(String(projectionVersion));
  return createHash("sha256").update(source, "utf8").digest("hex");
}

export function newMentionEntityIdFor(
  segmentId: string,
  sourceFingerprint: string,
  mentionId: string,
  claim: Pick<
    ExtractedClaim,
    "subject" | "predicate" | "object_entity" | "object_value"
  >,
): string {
  if ((claim.object_entity === null) === (claim.object_value === null)) {
    throw new ExtractionValidationError(
      "new mention identity requires exactly one entity object or literal value",
    );
  }
  const objectKind = claim.object_entity === null ? "literal" : "entity";
  const objectText = claim.object_entity ?? claim.object_value!;
  return uuidv5(
    "reflection-new-mention-v1:" +
      utf8Frame(segmentId) +
      utf8Frame(sourceFingerprint) +
      utf8Frame(mentionId) +
      utf8Frame(claim.subject) +
      utf8Frame(claim.predicate) +
      utf8Frame(objectKind) +
      utf8Frame(objectText),
    ENTITY_NAMESPACE,
  );
}

export function claimIdFor(segmentId: string, index: number): string {
  return uuidv5(`${segmentId}\0${index}`, CLAIM_NAMESPACE);
}

export function equivalenceKey(
  subjectId: string,
  predicate: string,
  object: { objectEntityId: string | null; objectValue: string | null },
): string {
  if ((object.objectEntityId === null) === (object.objectValue === null)) {
    throw new ExtractionValidationError(
      "equivalence key requires exactly one entity object or literal value",
    );
  }
  const objectKey =
    object.objectEntityId !== null
      ? `entity:${object.objectEntityId}`
      : `literal:${normalizeName(object.objectValue ?? "")}`;
  return createHash("sha256")
    .update(`${subjectId}\0${normalizeName(predicate)}\0${objectKey}`, "utf8")
    .digest("hex");
}

export interface EntityCandidate {
  id: string;
  canonicalName: string;
  description: string;
  aliases: readonly string[];
}

export interface MentionContext {
  mentionId: string;
  role: string;
  text: string;
  supportingClaim: string;
  candidates: readonly EntityCandidate[];
}

export interface PreparedEntity {
  id: string;
  canonicalName: string;
  normalizedName: string;
  description: string;
  aliases: readonly string[];
  embedding: readonly number[] | null;
  isNew: boolean;
}

export interface PreparedClaim {
  id: string;
  subject: string;
  subjectEntityId: string;
  predicate: string;
  confidence: number;
  objectEntity: string | null;
  objectEntityId: string | null;
  objectValue: string | null;
  equivalenceKey: string;
  embedding: readonly number[];
}

export interface PreparedSegment {
  id: string;
  sessionId: string;
  startUserMessageId: string;
  endUserMessageId: string;
  sourceBoundaryVersion: 1 | 2;
  startSourceMessageId: string | null;
  endSourceMessageId: string | null;
  summary: string;
  entities: readonly PreparedEntity[];
  claims: readonly PreparedClaim[];
  projectionVersion: number;
}

export interface RecallCandidate {
  subject: string;
  subjectEntityId: string;
  predicate: string;
  confidence: number;
  objectEntity: string | null;
  objectEntityId: string | null;
  objectValue: string | null;
  equivalenceKey: string;
  segmentId: string;
  similarity: number;
  seedSimilarity: number | null;
  isDirect: boolean;
}

export interface ClaimSupport {
  segmentIds: string[];
  supportCount: number;
  sessionCount: number;
}

export function unionCandidates(
  trigram: readonly EntityCandidate[],
  vector: readonly EntityCandidate[],
): EntityCandidate[] {
  const byId = new Map<string, EntityCandidate>();
  for (const candidate of [...trigram.slice(0, 5), ...vector.slice(0, 5)]) {
    if (!byId.has(candidate.id)) byId.set(candidate.id, candidate);
  }
  return [...byId.values()];
}

export interface ValidatedClaimDecision {
  index: number;
  claim: ExtractedClaim;
}

export interface ValidatedMentionResolution {
  context: MentionContext;
  resolution: Resolution;
  selectedCandidate: EntityCandidate | null;
}

export interface ValidatedResolutionPlan {
  keptClaims: readonly ValidatedClaimDecision[];
  mentions: readonly ValidatedMentionResolution[];
}

function validateClaimDecisions(
  proposed: readonly ExtractedClaim[],
  result: ResolutionResult,
): ValidatedClaimDecision[] {
  const expected = new Set(proposed.map((_, index) => `c${index}`));
  const actual = new Map(
    result.claims.map((decision) => [decision.claim_id, decision]),
  );
  if (actual.size !== result.claims.length) {
    throw new ExtractionValidationError(
      "claim triage returned duplicate claim IDs",
    );
  }
  if (
    actual.size !== expected.size ||
    [...actual.keys()].some((id) => !expected.has(id))
  ) {
    throw new ExtractionValidationError(
      "claim triage must return every proposed claim exactly once",
    );
  }
  return proposed.flatMap((claim, index) =>
    actual.get(`c${index}`)!.action === "keep" ? [{ index, claim }] : [],
  );
}

export function validateResolutionResult(
  proposed: readonly ExtractedClaim[],
  contexts: readonly MentionContext[],
  result: ResolutionResult,
): ValidatedResolutionPlan {
  const keptClaims = validateClaimDecisions(proposed, result);
  const keptMentionIds = new Set<string>();
  for (const { index, claim } of keptClaims) {
    keptMentionIds.add(`c${index}.subject`);
    if (claim.object_entity !== null) keptMentionIds.add(`c${index}.object`);
  }
  const keptContexts = contexts.filter((context) =>
    keptMentionIds.has(context.mentionId),
  );
  const expected = new Map(
    keptContexts.map((context) => [context.mentionId, context]),
  );
  const contextIndexes = new Map(
    keptContexts.map((context, index) => [context.mentionId, index]),
  );
  const actual = new Map(
    result.resolutions.map((resolution) => [resolution.mention_id, resolution]),
  );
  if (actual.size !== result.resolutions.length) {
    throw new ExtractionValidationError(
      "entity resolution returned duplicate mention IDs",
    );
  }
  if (
    actual.size !== expected.size ||
    [...actual.keys()].some((mentionId) => !expected.has(mentionId))
  ) {
    throw new ExtractionValidationError(
      "entity resolution must return every mention exactly once",
    );
  }
  const mentions: ValidatedMentionResolution[] = [];
  for (const context of keptContexts) {
    const resolution = actual.get(context.mentionId)!;
    const selectedCandidate =
      resolution.candidate_entity_id === null
        ? null
        : (context.candidates.find(
            (candidate) => candidate.id === resolution.candidate_entity_id,
          ) ?? null);
    if (resolution.candidate_entity_id !== null && selectedCandidate === null) {
      throw new ExtractionValidationError(
        `entity resolution selected an unknown candidate for ${context.mentionId}`,
      );
    }
    if (resolution.same_new_entity_as !== null) {
      const groupedContext = expected.get(resolution.same_new_entity_as);
      const groupedResolution = actual.get(resolution.same_new_entity_as);
      if (
        resolution.candidate_entity_id !== null ||
        groupedContext === undefined ||
        groupedResolution === undefined ||
        groupedResolution.candidate_entity_id !== null ||
        groupedResolution.same_new_entity_as !== null ||
        contextIndexes.get(resolution.same_new_entity_as)! >=
          contextIndexes.get(context.mentionId)! ||
        normalizeName(groupedContext.text) !== normalizeName(context.text)
      ) {
        throw new ExtractionValidationError(
          `entity resolution returned an invalid new-entity group for ${context.mentionId}`,
        );
      }
    }
    mentions.push({ context, resolution, selectedCandidate });
  }
  return { keptClaims, mentions };
}

function clamp(value: number): number {
  return Number.isNaN(value) ? 0 : Math.min(1, Math.max(0, value));
}

export function rankAndGroupClaims(
  candidates: Iterable<RecallCandidate>,
  supportByKey: ReadonlyMap<string, ClaimSupport>,
  limit = 20,
): SearchClaim[] {
  const grouped = new Map<
    string,
    { candidate: RecallCandidate; score: number }
  >();
  for (const candidate of candidates) {
    const similarity = clamp(candidate.similarity);
    const confidence = clamp(candidate.confidence);
    const support = supportByKey.get(candidate.equivalenceKey);
    const sessionCount = support?.sessionCount ?? 1;
    const supportBoost = Math.min(0.1, 0.02 * Math.max(0, sessionCount - 1));
    const score = candidate.isDirect
      ? 1 + 0.8 * similarity + 0.2 * confidence + supportBoost
      : 0.8 *
          (0.5 * similarity +
            0.3 * clamp(candidate.seedSimilarity ?? 0) +
            0.2 * confidence) +
        supportBoost;
    const current = grouped.get(candidate.equivalenceKey);
    if (!current || score > current.score)
      grouped.set(candidate.equivalenceKey, { candidate, score });
  }
  return [...grouped.values()]
    .sort((left, right) => right.score - left.score)
    .slice(0, limit)
    .map(({ candidate, score }) => {
      const support = supportByKey.get(candidate.equivalenceKey);
      return {
        subject: candidate.subject,
        subject_entity_id: candidate.subjectEntityId,
        predicate: candidate.predicate,
        confidence: candidate.confidence,
        object_entity: candidate.objectEntity,
        object_entity_id: candidate.objectEntityId,
        object_value: candidate.objectValue,
        segment_ids: support?.segmentIds ?? [candidate.segmentId],
        support_count: support?.supportCount ?? 1,
        session_count: support?.sessionCount ?? 1,
        score,
      };
    });
}
